test/services/version-service.ts

```diff
@@ -0,0 +1,194 @@
+		const testInjector = new Yok();
+		testInjector.register("logger", {
+			trace: (formatStr?: any, ...args: any[]): void => (undefined)
+		});
+		testInjector.register("fs", {});
+		testInjector.register("httpClient", {});
+		testInjector.register("projectDataService", {});
+		return testInjector;
+	};
+
+	let isHttpRequestCalled = false;
+	afterEach(() => {
+		process.env.TNS_CLI_CLOUD_VERSION = TNS_CLI_CLOUD_VERSION_ORIGINAL;
+	});
+
+	beforeEach(() => {
+		isHttpRequestCalled = false;
+		process.env.TNS_CLI_CLOUD_VERSION = "";
+	});
+
+	describe("getCliVersion", () => {
+		const cliVersion = "5.0.0";
+		it("returns process.env.TNS_CLI_CLOUD_VERSION when it is set", async () => {
+			const testInjector = createTestInjector();
+			process.env.TNS_CLI_CLOUD_VERSION = cliVersion;
+
+			const versionService = testInjector.resolve<IVersionService>(VersionService);
+			const actualCliVersion = await versionService.getCliVersion(null);
+			assert.equal(actualCliVersion, cliVersion);
+		});
+
+		const testData: any[] = [
+			{
+				runtimeVersion: "5.0.0",
+				expectedCliVersion: "5.0.2"
+			},
+			{
+				runtimeVersion: "5.0.2",
+				expectedCliVersion: "5.0.2"
+			},
+			{
+				runtimeVersion: "5.0.5",
+				expectedCliVersion: "5.0.2"
+			},
+			{
+				runtimeVersion: "5.1.5",
+				expectedCliVersion: "5.1.0"
+			},
+			{
+				runtimeVersion: "4.1.0",
+				expectedCliVersion: "4.1.0"
+			},
+			{
+				runtimeVersion: "4.1.8",
+				expectedCliVersion: "4.1.0"
+			}
+		];
+
+		const mockHttpRequest = (testInjector: IInjector, body: IDictionary<any>, error?: Error) => {
+			const httpClient = testInjector.resolve<Server.IHttpClient>("httpClient");
+
+			httpClient.httpRequest = async (url: string): Promise<any> => {
+				isHttpRequestCalled = true;
+				assert.equal(url, "http://registry.npmjs.org/nativescript");
+				if (error) {
+					throw error;
+				}
+
+				return { body: JSON.stringify(body) };
+
+			};
+		};
+
+		_.each(testData, ({ runtimeVersion, expectedCliVersion }) => {
+			it(`returns correct CLI version, based on runtime version, input: ${runtimeVersion}, expectedOutput: ${expectedCliVersion}`, async () => {
+				const testInjector = createTestInjector();
+				const body = {
+					versions: {
+						"1.0.0": {},
+						"5.0.0": {},
+						"5.0.1": {},
+						"5.0.2": {},
+						"5.1.0": {},
+						"6.0.0": {},
+					}
+				};
+
+				mockHttpRequest(testInjector, body);
+				const versionService = testInjector.resolve<IVersionService>(VersionService);
+				const actualCliVersion = await versionService.getCliVersion(runtimeVersion);
+				assert.equal(actualCliVersion, expectedCliVersion);
+				assert.isTrue(isHttpRequestCalled, "An http request to registry.npmjs.org should have been made in order to get data for CLI.");
+			});
+		});
+
+		it("fails when unable to determine CLI version as runtime version is not valid semver version", async () => {
+			const testInjector = createTestInjector();
+			const body = {
```

Comment:
Why bother to create an unused object - you can just call
`mockHttpRequest(testInjector, {});`
down below

Reply:
Just to ensure there's some data.